src/openrct2-ui/drawing/engines/opengl/TextureCache.cpp

```diff
@@ -128,36 +87,61 @@ void TextureCache::CreateAtlasesTexture()
         glGetIntegerv(GL_MAX_ARRAY_TEXTURE_LAYERS, &_atlasesTextureIndicesLimit);
         if (_atlasesTextureDimensions < _atlasesTextureIndicesLimit) _atlasesTextureIndicesLimit = _atlasesTextureDimensions;
 
-        AllocateAtlasesTexture();
-
-        _atlasesTextureInitialised = true;
+        glGenTextures(1, &_atlasesTexture);
+        glBindTexture(GL_TEXTURE_2D_ARRAY, _atlasesTexture);
+        glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+        glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+        glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
+        
+        glGenTextures(1, &_paletteTexture);
+        glBindTexture(GL_TEXTURE_RECTANGLE, _paletteTexture);
+        glTexParameteri(GL_TEXTURE_RECTANGLE, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+        glTexParameteri(GL_TEXTURE_RECTANGLE, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+        glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
+        GeneratePaletteTexture();
+
+        _initialized = true;
         _atlasesTextureIndices = 0;
     }
 }
 
-void TextureCache::AllocateAtlasesTexture()
+void TextureCache::GeneratePaletteTexture()
 {
-    // Create an array texture to hold all of the atlases
-    glGenTextures(1, &_atlasesTexture);
-    glBindTexture(GL_TEXTURE_2D_ARRAY, _atlasesTexture);
-    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
-    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
-    glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
+    rct_drawpixelinfo dpi = CreateDPI(256, PALETTE_TO_G1_OFFSET_COUNT + 5);
+    
+    // Init no-op palette
+    for (int i=0; i < 256; ++i)
```

Comment:
Why init to 0->255? Is there a reason 0 cant be used?

Reply:
The idea here is to allow for a no-op without a branch in the shader. If you look at the ApplyTransparency shader, you can see it used there. Our use of branches in the DrawImage/DrawRect shader [is almost Undefined Behavior](https://www.khronos.org/opengl/wiki/Sampler_(GLSL)#Non-uniform_flow_control), so I added a way for producing the input value as output without requiring a branch. This isn't 100% necessary, but having it is still useful.

Plus with the water palette being 5 pixels high, adding a few more rows doesn't increase the complexity of mapping from palette index to y value all that much.